testsuite/forge.py

```diff
@@ -0,0 +1,815 @@
+import json
+import multiprocessing
+import os
+import pwd
+import re
+import resource
+import subprocess
+import sys
+import tempfile
+import textwrap
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+from enum import Enum
+from typing import Any, Callable, Generator, List, Optional, Sequence, Tuple, Union
+
+
+@dataclass
+class RunResult:
+    exit_code: int
+    output: bytes
+
+    def unwrap(self) -> bytes:
+        if self.exit_code != 0:
+            raise Exception(self.output.decode("utf-8"))
+        return self.output
+
+
+class Shell:
+    def run(self, command: Sequence[str], stream_output: bool = False) -> RunResult:
+        raise NotImplementedError
+
+
+@dataclass
+class LocalShell(Shell):
+    verbose: bool = False
+
+    def run(self, command: Sequence[str], stream_output: bool = False) -> RunResult:
+        # Write to a temp file, stream to stdout
+        tmpname = tempfile.mkstemp()[1]
+        with open(tmpname, 'wb') as writer, open(tmpname, 'rb') as reader:
+            if self.verbose:
+                print(f"+ {' '.join(command)}")
+            process = subprocess.Popen(command, stdout=writer, stderr=writer)
+            output = b""
+            while process.poll() is None:
+                chunk = reader.read()
+                output += chunk
+                if stream_output:
+                    sys.stdout.write(chunk.decode("utf-8"))
+                time.sleep(0.1)
+            output += reader.read()
+        return RunResult(process.returncode, output)
+
+
+class FakeShell(Shell):
+    def run(self, command: Sequence[str], stream_output: bool = False) -> RunResult:
+        return RunResult(0, b'output')
+
+
+def install_dependency(dependency: str) -> None:
+    print(f"{dependency} is not currently installed")
+    answer = os.getenv("FORGE_INSTALL_DEPENDENCIES")
+    if not answer:
+        answer = input("Would you like to install it now? (y/n) ").strip().lower()
+    if answer in ("y", "yes", "yeet", "yessir", "si"):
+        shell = LocalShell(True)
+        shell.run(["pip3", "install", dependency], stream_output=True).unwrap()
+    else:
+        print(f"Please install click (pip install {dependency}) and try again")
+        exit(1)
+
+
+try:
+    import click
+except ImportError:
+    install_dependency("click")
+    import click
+
+try:
+    import psutil
+except ImportError:
+    install_dependency("psutil")
+    import psutil
+
+
+def get_current_user() -> str:
+    return pwd.getpwuid(os.getuid())[0]
+
+
+def get_utc_timestamp(dt: datetime) -> str:
+    return dt.strftime("%Y-%m-%dT%H:%M:%S.000Z")
+
+
+@click.group()
+def main() -> None:
+    # Check that the current directory is the root of the repository.
+    if not os.path.exists('.git'):
+        print('This script must be run from the root of the repository.')
```

Comment:
Same comment here with printing to stderr. Same thing for anywhere else we print and then exit.